Adopt more smart pointers in the Network Process

csaavedra opened the pull request:
#### 8d968dcd45f7ea6c40127b43909f02d9a788a6e0
<pre>
Adopt more smart pointers in the Network Process
<a href="https://bugs.webkit.org/show_bug.cgi?id=286681">https://bugs.webkit.org/show_bug.cgi?id=286681</a>

Reviewed by Chris Dumez.

Many fixes to how &apos;this&apos; objects are captured in lambdas,
that reduce the number of Safer CPP warnings.

* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::processStatisticsAndDataRecords):
(WebKit::ResourceLoadStatisticsStore::scheduleStatisticsProcessingRequestIfNecessary):
(WebKit::ResourceLoadStatisticsStore::debugBroadcastConsoleMessage):
(WebKit::ResourceLoadStatisticsStore::grantStorageAccess):
(WebKit::ResourceLoadStatisticsStore::dumpResourceLoadStatistics):
* Source/WebKit/NetworkProcess/Downloads/Download.cpp:
(WebKit::Download::cancel):
(WebKit::Download::didFinish):
* Source/WebKit/NetworkProcess/NetworkDataTask.cpp:
(WebKit::NetworkDataTask::scheduleFailure):
* Source/WebKit/NetworkProcess/NetworkLoad.cpp:
(WebKit::NetworkLoad::willPerformHTTPRedirection):
* Source/WebKit/NetworkProcess/NetworkLoadChecker.cpp:
(WebKit::NetworkLoadChecker::validateResponse):
(WebKit::NetworkLoadChecker::checkCORSRequestWithPreflight):
(WebKit::NetworkLoadChecker::processContentRuleListsForLoad):
* Source/WebKit/NetworkProcess/PingLoad.cpp:
(WebKit::PingLoad::initialize):
(WebKit::PingLoad::willPerformHTTPRedirection):
* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementManager.cpp:
(WebKit::PrivateClickMeasurementManager::storeUnattributed):
(WebKit::PrivateClickMeasurementManager::getTokenPublicKey):
(WebKit::PrivateClickMeasurementManager::getSignedUnlinkableTokenForSource):
(WebKit::PrivateClickMeasurementManager::getSignedUnlinkableTokenForDestination):
(WebKit::PrivateClickMeasurementManager::handleAttribution):
(WebKit::PrivateClickMeasurementManager::attribute):
(WebKit::PrivateClickMeasurementManager::fireConversionRequest):
(WebKit::PrivateClickMeasurementManager::fireConversionRequestImpl):
(WebKit::PrivateClickMeasurementManager::firePendingAttributionRequests):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWRegistrationStore.cpp:
(WebKit::WebSWRegistrationStore::updateToStorage):
* Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
(WebKit::NetworkCache::Cache::startAsyncRevalidationIfNeeded):
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
(WebKit::NetworkCache::Storage::traverseWithinRootPath):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
(WebKit::NetworkDataTaskCocoa::resume):
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::SessionSet::isolatedSession):
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
(WebKit::CacheStorageManager::initializeCacheSize):
* Source/WebKit/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:

Canonical link: <a href="https://commits.webkit.org/289805@main">https://commits.webkit.org/289805@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e6d1d648f126f3a6b8fc1b40407b290cf6e5568e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88028 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7544 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42426 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92976 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38778 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7925 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15722 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67959 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25689 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91030 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6072 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79644 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48320 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5847 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37885 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76240 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34936 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94824 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15196 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11174 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76808 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15451 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75500 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76065 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18703 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20438 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18847 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8206 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15214 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20515 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14956 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18401 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16738 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -2160,7 +2163,7 @@ void ResourceLoadStatisticsStore::dumpResourceLoadStatistics(CompletionHandler<v
     ASSERT(!RunLoop::isMain());
     if (m_dataRecordsBeingRemoved) {
-        m_dataRecordRemovalCompletionHandlers.append([this, completionHandler = WTFMove(completionHandler)]() mutable {
+        m_dataRecordRemovalCompletionHandlers.append([this, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler)]() mutable {
```
This looks like a ref-cycle to me and thus leaking. This needs to use `weakThis`.

cdumez commented on Source/WebKit/NetworkProcess/NetworkLoadChecker.cpp:
```diff
@@ -214,7 +214,7 @@ ResourceError NetworkLoadChecker::validateResponse(const ResourceRequest& reques
         return badResponseHeadersError(request.url());
 
-    auto scope = makeScopeExit([&] {
+    auto scope = makeScopeExit([this, protectedThis = Ref { *this }, &response] {
```
@rniwa don't we want to add `NOESCAPE` for the parameter to makeScopeExit instead?

csaavedra replied:
I see this pattern in a few places, so not sure what to do.

```
Source/WebKit/WebProcess/WebPage/WebFrame.cpp:    return makeScopeExit<Function<void()>>([protectedThis = Ref { *this }] {
Source/WebKit/UIProcess/WebProcessProxy.cpp:    auto scope = makeScopeExit([this, protectedThis = Ref { *this }]() {
Source/WebKit/WebProcess/WebPage/WebHistoryItemClient.cpp:    return makeScopeExit(CompletionHandler<void()> { [this, protectedThis = Ref { *this }] {
Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:    auto scopeExit = WTF::makeScopeExit([protectedThis = Ref { *this }] {
```

rniwa replied:
I don't think we can add `NOESCAPE` for `makeScopeExit` since the point of `makeScopeExit` to create an object which runs the function in the destructor, meaning that it stores the lambda.

cdumez commented on Source/WebKit/NetworkProcess/NetworkLoadChecker.cpp:
```diff
@@ -490,7 +490,7 @@ void NetworkLoadChecker::checkCORSRequestWithPreflight(ResourceRequest&& request
         request.hasHTTPHeaderField(HTTPHeaderName::SecFetchSite)
     };
-    m_corsPreflightChecker = NetworkCORSPreflightChecker::create(m_networkProcess.get(), m_networkResourceLoader.get(), WTFMove(parameters), m_shouldCaptureExtraNetworkLoadMetrics, [this, request = WTFMove(request), handler = WTFMove(handler), isRedirected = isRedirected()](auto&& error) mutable {
+    m_corsPreflightChecker = NetworkCORSPreflightChecker::create(m_networkProcess.get(), m_networkResourceLoader.get(), WTFMove(parameters), m_shouldCaptureExtraNetworkLoadMetrics, [this, protectedThis = Ref { *this }, request = WTFMove(request), handler = WTFMove(handler), isRedirected = isRedirected()](auto&& error) mutable {
```
I would feel better if we used weakThis here given that `m_corsPreflightChecker` is a data member. I'm worried about ref cycle here, though I didn't look super deeply.

cdumez commented on Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
```diff
@@ -2040,7 +2040,7 @@ void didReceiveData(const WebCore::SharedBuffer& buffer)
         if (!m_connection)
             return;
-        buffer.forEachSegment([&](auto segment) {
+        buffer.forEachSegment([this, protectedThis = Ref { *this }](auto segment) {
```
I think we just need to mark the `forEachSegment()` parameter as `NOESCAPE` and we won't need this change. This lambda runs synchronously so it is safe.

csaavedra replied:
IIUC this is a change to `SharedBuffer::forEachSegment()` and probably also `SharedBuffer::forEachSegmentAsSharedBuffer()`, would you prefer if I make that a different patch? I think it would make sense.

csaavedra replied:
Done in https://github.com/WebKit/WebKit/pull/39813

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
```diff
@@ -381,7 +381,7 @@ void Cache::startAsyncRevalidationIfNeeded(const WebCore::ResourceRequest& reque
             return WeakHashSet<AsyncRevalidation>();
         });
-        Ref revalidation = AsyncRevalidation::create(*this, frameID, request, WTFMove(entry), isNavigatingToAppBoundDomain, allowPrivacyProxy, advancedPrivacyProtections, [this, key](auto result) {
+        Ref revalidation = AsyncRevalidation::create(*this, frameID, request, WTFMove(entry), isNavigatingToAppBoundDomain, allowPrivacyProxy, advancedPrivacyProtections, [this, protectedThis = Ref { *this }, key](auto result) {
```
Again very risky to capture `protectedThis` here given that this ends up being stored in a data member.

cdumez commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1775,7 +1778,7 @@ void ResourceLoadStatisticsStore::grantStorageAccess(SubFrameDomain&& subFrameDo
     auto transactionScope = beginTransactionIfNecessary();
 
-    auto addGrant = [&, frameID, pageID, promptWasShown, scope] (SubFrameDomain&& subFrameDomain, TopFrameDomain&& topFrameDomain, CanRequestStorageAccessWithoutUserInteraction canRequestStorageAccessWithoutUserInteraction, CompletionHandler<void(StorageAccessWasGranted)>&& completionHandler) mutable {
+    auto addGrant = [&, protectedThis = Ref { *this }, frameID, pageID, promptWasShown, scope] (SubFrameDomain&& subFrameDomain, TopFrameDomain&& topFrameDomain, CanRequestStorageAccessWithoutUserInteraction canRequestStorageAccessWithoutUserInteraction, CompletionHandler<void(StorageAccessWasGranted)>&& completionHandler) mutable {
```
We should drop the `&` here. This lambda gets passed to another thread and runs asynchronously. Capturing anything by referencing here wouldn't be safe.

cdumez commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -2160,8 +2163,10 @@ void ResourceLoadStatisticsStore::dumpResourceLoadStatistics(CompletionHandler<v
-            dumpResourceLoadStatistics(WTFMove(completionHandler));
+        m_dataRecordRemovalCompletionHandlers.append([weakThis = WeakPtr { *this }, completionHandler = WTFMove(completionHandler)]() mutable {
+            RefPtr protectedThis = weakThis.get();
+            if (protectedThis)
```
You still need to call the completion handler in the `else` case or we will crash (completion handlers must be called before destruction).

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
```diff
@@ -381,9 +381,11 @@ void Cache::startAsyncRevalidationIfNeeded(const WebCore::ResourceRequest& reque
-            ASSERT(m_pendingAsyncRevalidations.contains(key));
-            m_pendingAsyncRevalidations.remove(key);
+        Ref revalidation = AsyncRevalidation::create(*this, frameID, request, WTFMove(entry), isNavigatingToAppBoundDomain, allowPrivacyProxy, advancedPrivacyProtections, [weakThis = WeakPtr { *this }, key](auto result) {
+            if (weakThis) {
```
Why no `protectedThis` here? Seems a little risky to be using `weakThis` without protection.

csaavedra replied:
I assumed that because all uses of `weakThis` are trivial, it wouldn't matter, but I can change it if you prefer.

cdumez replied:
Yes, I think we just avoid any risk in modern code, especially when there is no performance concern.

cdumez reviewed (approved):
LGTM

csaavedra commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -716,11 +718,12 @@ void ResourceLoadStatisticsStore::didCreateNetworkProcess()
-        RunLoop::main().dispatch([&, weakThis = WeakPtr { *this }, source = crossThreadCopy(source), level = crossThreadCopy(level), message = crossThreadCopy(message)]() {
-            if (!weakThis)
+        RunLoop::main().dispatch([weakThis = WeakPtr { *this }, source = crossThreadCopy(source), level = crossThreadCopy(level), message = crossThreadCopy(message)]() {
+            RefPtr protectedThis = weakThis.get();
```
This is wrong because the dispatch moves the code execution to the main loop thread. I am fixing this.

RupinMittal commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -716,11 +718,9 @@ void ResourceLoadStatisticsStore::didCreateNetworkProcess()
-            debugBroadcastConsoleMessage(source, level, message);
+        RunLoop::main().dispatch([store = Ref { store() }, source = crossThreadCopy(source), level = crossThreadCopy(level), message = crossThreadCopy(message)]() {
+            if (auto *networkSession = store->networkSession())
+                networkSession->networkProcess().broadcastConsoleMessage(networkSession->sessionID(), source, level, message);
```
Could you explain this change? It seems like the right thing to do would be to just remove the & capture of `this` and use `protectedThis` instead of removing/refactoring the call to `debugBroadcastConsoleMessage`. Like this:
```
RunLoop::main().dispatch([weakThis = WeakPtr { *this }, source = crossThreadCopy(source), level = crossThreadCopy(level), message = crossThreadCopy(message)]() {
    if (RefPtr protectedThis = weakThis.get())
        protectedThis->debugBroadcastConsoleMessage(source, level, message);
    });
```

csaavedra replied:
I explained above why that wouldn't work. I had this in a previous version of the patch, but `weakThis.get()` asserts because the lambda is executed in the main loop thread, and `WeakPtr` is not thread-safe. Instead of making `ResourceLoadStatisticsStore` inherit from `ThreadSafeWeakPtr` I relied on what is done for other `RunLoop::main().dispatch()` calls throughout the class, that is, just capture a reference to the `WebResourceLoadStatisticsStore` member, which already implements `ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr`.

I can go the other way tho, if you think that's better.

RupinMittal replied:
Ah I must have missed your explanation above! That does make sense. Thank you for clarifying. Keeping it as-is seems good.

RupinMittal commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1775,7 +1775,7 @@ void ResourceLoadStatisticsStore::grantStorageAccess(SubFrameDomain&& subFrameDo
     auto transactionScope = beginTransactionIfNecessary();
 
-    auto addGrant = [&, frameID, pageID, promptWasShown, scope] (SubFrameDomain&& subFrameDomain, TopFrameDomain&& topFrameDomain, CanRequestStorageAccessWithoutUserInteraction canRequestStorageAccessWithoutUserInteraction, CompletionHandler<void(StorageAccessWasGranted)>&& completionHandler) mutable {
+    auto addGrant = [this, protectedThis = Ref { *this }, frameID, pageID, promptWasShown, scope] (SubFrameDomain&& subFrameDomain, TopFrameDomain&& topFrameDomain, CanRequestStorageAccessWithoutUserInteraction canRequestStorageAccessWithoutUserInteraction, CompletionHandler<void(StorageAccessWasGranted)>&& completionHandler) mutable {
```
It looks like `addGrant` will be executed asynchronously (it's captured by a lambda given to `RunLoop::dispatch()` below). So we probably want to capture `weakThis` instead of `protectedThis`.

csaavedra committed:
Adopt more smart pointers in the Network Process
https://bugs.webkit.org/show_bug.cgi?id=286681

Reviewed by Chris Dumez.

Many fixes to how 'this' objects are captured in lambdas,
that reduce the number of Safer CPP warnings.

* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::processStatisticsAndDataRecords):
(WebKit::ResourceLoadStatisticsStore::scheduleStatisticsProcessingRequestIfNecessary):
(WebKit::ResourceLoadStatisticsStore::debugBroadcastConsoleMessage):
(WebKit::ResourceLoadStatisticsStore::grantStorageAccess):
(WebKit::ResourceLoadStatisticsStore::dumpResourceLoadStatistics):
* Source/WebKit/NetworkProcess/Downloads/Download.cpp:
(WebKit::Download::cancel):
(WebKit::Download::didFinish):
* Source/WebKit/NetworkProcess/NetworkDataTask.cpp:
(WebKit::NetworkDataTask::scheduleFailure):
* Source/WebKit/NetworkProcess/NetworkLoad.cpp:
(WebKit::NetworkLoad::willPerformHTTPRedirection):
* Source/WebKit/NetworkProcess/NetworkLoadChecker.cpp:
(WebKit::NetworkLoadChecker::validateResponse):
(WebKit::NetworkLoadChecker::checkCORSRequestWithPreflight):
(WebKit::NetworkLoadChecker::processContentRuleListsForLoad):
* Source/WebKit/NetworkProcess/PingLoad.cpp:
(WebKit::PingLoad::initialize):
(WebKit::PingLoad::willPerformHTTPRedirection):
* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementManager.cpp:
(WebKit::PrivateClickMeasurementManager::storeUnattributed):
(WebKit::PrivateClickMeasurementManager::getTokenPublicKey):
(WebKit::PrivateClickMeasurementManager::getSignedUnlinkableTokenForSource):
(WebKit::PrivateClickMeasurementManager::getSignedUnlinkableTokenForDestination):
(WebKit::PrivateClickMeasurementManager::handleAttribution):
(WebKit::PrivateClickMeasurementManager::attribute):
(WebKit::PrivateClickMeasurementManager::fireConversionRequest):
(WebKit::PrivateClickMeasurementManager::fireConversionRequestImpl):
(WebKit::PrivateClickMeasurementManager::firePendingAttributionRequests):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWRegistrationStore.cpp:
(WebKit::WebSWRegistrationStore::updateToStorage):
* Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
(WebKit::NetworkCache::Cache::startAsyncRevalidationIfNeeded):
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
(WebKit::NetworkCache::Storage::traverseWithinRootPath):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
(WebKit::NetworkDataTaskCocoa::resume):
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::SessionSet::isolatedSession):
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
(WebKit::CacheStorageManager::initializeCacheSize):
* Source/WebKit/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:

Canonical link: https://commits.webkit.org/289805@main
